src/Adapter/Order/OrderProductUpdater.php

```diff
@@ -0,0 +1,545 @@
+    {
+        Context::getContext()->cart = $cart;
+        CartRule::autoAddToCart();
+        CartRule::autoRemoveFromCart();
+
+        $computingPrecision = $this->getPrecisionFromCart($cart);
+        $newCartRules = $cart->getCartRules();
+        foreach ($order->getCartRules() as $orderCartRuleData) {
+            foreach ($newCartRules as $newCartRule) {
+                if ($newCartRule['id_cart_rule'] == $orderCartRuleData['id_cart_rule']) {
+                    // Cart rule is still in the cart no need to remove it, but we update it as the amount may have changed
+                    $cartRule = new CartRule($newCartRule['id_cart_rule']);
+
+                    $orderCartRule = new OrderCartRule($orderCartRuleData['id_order_cart_rule']);
+                    $orderCartRule->id_order = $order->id;
+                    $orderCartRule->name = $newCartRule['name'];
+                    $orderCartRule->value = Tools::ps_round($cartRule->getContextualValue(true), $computingPrecision);
+                    $orderCartRule->value_tax_excl = Tools::ps_round($cartRule->getContextualValue(false), $computingPrecision);
+                    $orderCartRule->save();
+                    continue 2;
+                }
+            }
+
+            // This one is no longer in the new cart rules so we delete it
+            $orderCartRule = new OrderCartRule($orderCartRuleData['id_order_cart_rule']);
+            if (!$orderCartRule->delete()) {
+                throw new OrderException('Could not delete order cart rule from database.');
+            }
+        }
+
+        // Finally add the new cart rules that are not in the Order
+        foreach ($newCartRules as $newCartRule) {
+            foreach ($order->getCartRules() as $orderCartRuleData) {
+                if ($newCartRule['id_cart_rule'] == $orderCartRuleData['id_cart_rule']) {
+                    // This cart rule is already present no need to add it
+                    continue 2;
+                }
+            }
+
+            // Add missing order cart rule
+            $cartRule = new CartRule($newCartRule['id_cart_rule']);
+
+            $orderCartRule = new OrderCartRule();
+            $orderCartRule->id_order = $order->id;
+            $orderCartRule->id_cart_rule = $newCartRule['id_cart_rule'];
+            $orderCartRule->id_order_invoice = $order->getInvoicesCollection()->getLast();
+            $orderCartRule->name = $newCartRule['name'];
+            $orderCartRule->value = Tools::ps_round($cartRule->getContextualValue(true), $computingPrecision);
+            $orderCartRule->value_tax_excl = Tools::ps_round($cartRule->getContextualValue(false), $computingPrecision);
+            $orderCartRule->save();
+        }
+    }
+
+    /**
+     * @param Cart $cart
+     * @param Order $order
+     *
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function recalculateCartRules(Cart $cart, Order $order): void
```

Comment:
I believe this one  is redundant with `updateOrderCartRules` where we already loop through cart rules and update OrderCartRule when they are still present

Reply:
done